src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,17 +105,170 @@ def pileupSizeTask(self):
             self.logger.exception(msg)
         return cmsDict, cusDict
 
+    def getPileupBlocks(self, doc, previousFraction, fraction):
+        """
+        Return list of blocks for increase/decrease scenarios of partial pileup
+        according to logic outlined in
+        https://gist.github.com/amaltaro/b4f9bafc0b58c10092a0735c635538b5#logic-for-increasingdecreasing-container-fraction
+        :return: list of blocks
+        """
+        if previousFraction < fraction:
+            return self.getIncreasingBlocks(doc, fraction)
+        return self.getDecreasingBlocks(doc, fraction)
+
+    def getIncreasingBlocks(self, doc, fraction):
+        """
+        Return list of blocks for increase scenarios of partial pileup
+        according to logic outlined in
+        https://gist.github.com/amaltaro/b4f9bafc0b58c10092a0735c635538b5#logic-for-increasingdecreasing-container-fraction
+        NOTE: in order to increaee pileup fraction we already should have
+        in place transition record since MSPilupeObj always created with container fraction 1
+
+        :param doc: pileup document
+        :param fraction: container fraction
+        :return: list of blocks
+        """
+        pname = doc['pileupName']
+        lastTransition = doc['transition'][-1]
+        cname = lastTransition['customDID']
+        totalBlocks = self.rucioClient.getBlocksInContainer(pname)
+        customBlocks = self.rucioClient.getBlocksInContainer(cname)
+        portion = math.ceil(fraction * len(totalBlocks))
+        blockList = customBlocks + [b for b in totalBlocks if b not in customBlocks]
+        self.logger.info("increase scenario: use %d blocks out of %d custom blocks from %s and %d from %s",
+                         len(blockList), len(customBlocks), cname, len(totalBlocks), pname)
+        return blockList[:portion]
+
+    def getDecreasingBlocks(self, doc, fraction):
+        """
+        Return list of blocks for decrease scenarios of partial pileup
+        according to logic outlined in
+        https://gist.github.com/amaltaro/b4f9bafc0b58c10092a0735c635538b5#logic-for-increasingdecreasing-container-fraction
+
+        :param doc: pileup document
+        :param fraction: container fraction
+        :return: list of blocks
+        """
+        pname = doc['pileupName']
+        cname = doc['customName']
+        totalBlocks = self.rucioClient.getBlocksInContainer(pname)
+        portion = math.ceil(fraction * len(totalBlocks))
+        if cname == '':
+            self.logger.info("decrease scenario: use %d blocks out of pileup %s with %d blocks",
+                             len(totalBlocks[:portion]), pname, len(totalBlocks))
```

Comment:
Isn't it easier - and faster - to provide `portion` as the first argument to this record?

Reply:
yes, making necessary changes.